### DISCUSSION TITLE
What is the different(and stand for) for cycles:Pu(self) and cycles:Pu(incl.) in the caller/callee panel?

### DISCUSSION THREAD

**Discussion by zw963** [KDAB/hotspot#681] [Q&A] [answered]
Sorry for ask question here, because i am a newbie for use perf profile, and i search on all search engine for find it out, but no any clue.

I generate profile use following command:

```
$: perf record --call-graph dwarf -- bin/tartrazine ../crycco/src/crycco.cr swapoff
```

Then, use hotspot open the `perf.data`, like following:

```
$: hotspot perf.data
```

I get a window like following screenshot.

![image](https://github.com/user-attachments/assets/043cce24-4e30-4d61-bb9d-d737ea1113f6)

I don't know what `cycles:Pu(self)` and `cycles:Pu(incl.)` stand for in above caller/callee panel, sure i can't understand why others functions those two column value almost same, but, for `char_bytesize_at` function, why `self` is 8.83%, and `incl.` is 88.9%?

is there a link for keyword for study knowledge of this?  could you plese provide some guidance? Thank you.
Thanks

**Comment by milianw** [comment, marked as answer]
`self` mean samples that point directly at code from this function, and `incl.` means cost of this function including all other functions called by it.

the `:P` modifier is the "as precise as possible" modifier for x86/intel PMU - I think the kernel documentations will talk about this somwhere, maybe in `man perf record`. the `:u` modifier means "only userspace" - i.e. excluding cost in kernel space.

does this answer your questions?

**Reply by zw963** [reply]
`man perf record`  indeed many useful information, I will definitely read it thoroughly.

Thanks a lot.

**Comment by zw963** [comment]
One more question, is there a way to get a function invoke count? 

e.g. above `char_bytesize_at`, i want know this function took a lot of time indeed when invoke once, or it has been called very many times.

It that possible for hotspot?

**Reply by milianw** [reply]
No, a sampling profiler does not know what happens inbetween samples - it could be many repeated function calls or just one long running one. But you can use trace points for this purpose, e.g. via `perf probe` (also see its manpage on `uprobe`)

**Reply by zw963** [reply]
Thanks for help, I am learn a lot since asked question here, helped with GPT4o too.
